[Ong Zheng Lin] iP

Ongzl opened the pull request:
# DukeManager <h1>
To-do list manager.
* Text-based
* Easy to use

All you need to do is:
1. download the JAR file [here](https://github.com/Ongzl/ip/releases/tag/Level-9)
2. double click main.jar
3. run it and start keying in your tasks

totally **FREE**

Features:
- [x] Simple UI
- [x] Remembers to-do tasks
- [ ] Give reminders

Here's the main method:

    public static void main(String[] args) {
        new Duke().run();
    }

Commit:
Add Gradle support

Ongzl committed:
Level-1, echoing input

Ongzl committed:
Level-2, added list class for storing and showing list of activity. changed echoing input to adding input into the list and a list cmd to show the content of the list.

Ongzl committed:
Level-2, changed login function from if-else statements into switch statement and partially implemented level-3 checkboxes

Ongzl committed:
Level-3, implemented scanning done (number)

Ongzl committed:
Level-4, Added todo, event, deadline. replaced task in duke with todo. Added checks for done out of range. Added toString for task for better inheritance. Added different add tasks behavior to list

Ongzl committed:
A-TextUiTesting, added content to input.txt, changed EXPECTED.TXT to comply with input.txt.

Ongzl committed:
Level-5. Handled invalid input (eg todo    , event     ) and done with no number

Ongzl committed:
Level-6. Implemented delete task function as well as custom message when there is no take in the list

Ongzl committed:
Level-5A-Exceptions. Added DukeException.java as an umbrella exception to signify an unknown prompt

Ongzl committed:
Added extractInt function to Duke to make code neater, also added documentations and removed tostring() from addtodo/deadline/event

Ongzl committed:
Level-7 Add save function by implementing serializable interface for Task and add writeToFile and loadFromFile function in List

Ongzl committed:
Level-8, add LocalDate to Deadline but currently only accepts dd/mm/yyyy or dd-mm-yyyy as date format, also Deadline constructor now throws DukeException

Ongzl committed:
Merge branch 'branch-level-8'

Ongzl committed:
A-MoreOOP: add Parser, Storage, Ui for more OOP

Ongzl committed:
A-Packages: organize the files into packages util,task,exception

Ongzl committed:
A-JUnit, add DeadlineTest.java and ToDoTest.java for basic testing, updated exception handling in Deadline and List

Ongzl committed:
A-JavaDoc, add Javadocs to most methods and classes

Ongzl committed:
A-CodingStandard, ensure code conforms to coding standard

Ongzl committed:
Level-9, add find case to Parser.java as well as find method to List.java and getDescription method to Tasks to support find method

BryannYeap commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,31 @@
+
+    private final Ui ui;
+    private final List list;
+    private final Parser parser;
```
I believe constant names should be in all uppercase?

Ongzl replied:
yeap thanks for highlighting think the IDE changed it to final during formatting

BryannYeap commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,31 @@
+        new Duke().run();
+    }
+
+    /**
+     * Run the Duke program, prints out welcome statement and starts accepting input.
+     */
```
Javadoc headers should start with words in the form `Returns ...`, `Sends ...`, `Adds ...` (not `Return` or `Returnning` etc.)
This applies to your other javadoc headers in your other classes too! I believe I spotted some that do not follow this convention. 😬

felissaf replied:
I agree with @BryannYeap about javadocs. Most of your javadocs headers have the same problem.:smile:

Ongzl replied:
haha yea thanks for pointing out

Agentum07 replied:
Seconded.

BryannYeap commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,39 @@
+package duke.task;
+
+import duke.exception.DukeException;
+
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
```
I don't think its compulsory but this is the import order used in AB3:
https://se-education.org/guides/tutorials/intellijCodeStyle.html#tweak-import-order

Damith mentioned in the forum that we could follow this for iP as well. Just a consideration!

Ongzl replied:
nice thanks will check it out

BryannYeap commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done duke.task with X
```
This code should be self explanatory enough to be able to omit this comment? 🤔

Ongzl replied:
haha true

BryannYeap commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,99 @@
+        this.list = list;
+    }
+
+    // extract int from done and delete prompts
+    public static int extractInt(String input) throws DukeException {
+        String[] parts = input.split(" ");//split along the whitespace to get the integer
+        if (parts.length <= 1) {// checking for incomplete prompts
+            throw new DukeException();
+        }
+        String numStr = parts[1];
+        return Integer.valueOf(numStr);
+    }
```
I believe public non-trivial methods like these (as well as other methods in this class and other classes) should be documented by Javadocs. 😧

Ongzl replied:
yea updated thanks

Agentum07 replied:
I think you should come up with a more comprehensive javadoc for this method.
Use the `@param` and `@return` tags to explain what your input is and what you return

BryannYeap commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,99 @@
+                    continue;
+                }
+
+                switch (input) {
+                case "bye":
+                case "Bye":
+                case "BYE":
+                    System.out.println("Bye. Hope to see you again soon!");
```
Perhaps you could consider refactoring the code this way?
```java
switch (input.toLowerCase()) {
    case "bye" : 
        System.out.println("Bye. Hope to see you again soon!");
```

BryannYeap commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,99 @@
+        return Integer.valueOf(numStr);
+    }
+
+    public void execute() {
```
This method / while loop might be too long. At [this code quality link](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#structure-code-logically), they stated the following:
> Be wary when a method is longer than the computer screen, and take corrective action when it goes
> beyond 30 LOC (lines of code). The bigger the haystack, the harder it is to find a needle.

This is a little longer than 30 LOC 😅

BryannYeap commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,99 @@
+        return Integer.valueOf(numStr);
+    }
+
+    public void execute() {
```
This method / while loop might be too long. This [coding quality website](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-long-methods) stated the following:
> Be wary when a method is longer than the computer screen, and take corrective action when it goes
> beyond 30 LOC (lines of code). The bigger the haystack, the harder it is to find a needle.

This is a little longer than 30 LOC 😅

Xiaoyunnn replied:
I agree with @BryannYeap. Maybe you can extract the content in the `if` / `else if` / `else` block into separate methods to reduce the length of this method and make your code look cleaner?

BryannYeap commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,99 @@
+                    continue;
+                }
+
+                switch (input) {
+                case "bye":
+                case "Bye":
+                case "BYE":
+                    System.out.println("Bye. Hope to see you again soon!");
+                    shouldContinue = false;
+                    break;
```
Perhaps consider refactoring the code this way?
```java
                switch (input.toLowerCase()) {
                case "bye":
                    System.out.println("Bye. Hope to see you again soon!");
                    shouldContinue = false;
                    break;
```

Ongzl replied:
yea did not thought of this thanks

Agentum07 replied:
I agree with @BryannYeap, I was going to suggest the same haha!

BryannYeap commented on src/main/java/duke/util/List.java:
```diff
@@ -0,0 +1,149 @@
+     *
+     * @param input Keyword to search Tasks with.
+     */
+    public void find(String input) {
```
Shouldn't method names be a little more descriptive? Perhaps consider naming the method something that will tell us what it does (i.e. findTask)!

> A name is not just for differentiation; it should explain the named entity to the reader accurately and at a sufficient level of detail.

Ongzl replied:
nice catch thanks

BryannYeap commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,99 @@
+        this.list = list;
+    }
+
+    // extract int from done and delete prompts
```
I believe public non-trivial methods such as this (as well as the others in this class and other classes) should be well-documented by Javadocs. Do consider adding them!

BryannYeap commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,99 @@
+                    list.find(desc);
+                    input = sc.nextLine();
+                    continue;
+                } else if (input.startsWith("done ")) {
+                    int numInt = extractInt(input);
+                    list.setIndexDone(numInt);
+                    input = sc.nextLine();
+                    continue;
+                } else if (input.startsWith("delete ")) {
+                    int numInt = extractInt(input);
+                    list.deleteTask(numInt);
+                    input = sc.nextLine();
+                    continue;
+                } else if (input.startsWith("todo ")) {
+                    String task = input.replaceFirst("todo ", "");//remove to-do from input
+                    if (task.isBlank()) {//check for incomplete input
+                        throw new DukeException();
+                    }
+                    list.addToDo(task);
+                    input = sc.nextLine();
+                    continue;
+                } else if (input.startsWith("deadline ")) {
+                    String task = input.replaceFirst("deadline ", "");//remove deadline from input
+                    if (task.isBlank()) {//checking incomplete input
+                        throw new DukeException();
+                    }
+                    String[] parts = task.split("/by ");//split string along keyword /by
+                    list.addDeadline(parts[0], parts[1]);
+                    input = sc.nextLine();
+                    continue;
+                } else if (input.startsWith("event ")) {
+                    String task = input.replaceFirst("event ", "");//remove event from input
+                    if (task.isBlank()) {//checking incomplete input
+                        throw new DukeException();
+                    }
+                    String[] parts = task.split("/at");//split string along keyword /at
+                    list.addEvent(parts[0], parts[1]);
+                    input = sc.nextLine();
+                    continue;
+                }
```
Hopefully I am not misunderstanding your code (and correct me if I am!), but I believe that instead of the repeated use of
```java
input = sc.nextLine();
continue;
```

You could merely declare `String input = null;` or `String input;` at line 27, and then place the line 
`input = sc.nextLine()` at the start of the while loop between line 31 and 32. Then you can subsequently remove all lines of `continue` as well. 

Resultantly, in this program it will look at a singular `if` / `if else` block, complete an action, and then loop to get the next line of input.

The `switch` block below can then either be split into a separate method (to shorten this method as mentioned in my other comment), or simply be converted into more `if else` blocks.

Just a suggestion! 😸

Ongzl replied:
yea that would be alot cleaner thanks for the suggestion

BryannYeap reviewed (commented):
Just some coding quality and convention issues to solve first. Afterwards, it LGTM!

felissaf commented on src/main/java/duke/util/List.java:
```diff
@@ -0,0 +1,149 @@
+public class List {
+    //List that holds all the Tasks.
+    private final ArrayList<Task> list;
+    //Storage to load the list and write to list,
```
I believe there is a typo here, it should be a fullstop intead a comma at the end of the sentence.:smile:

Ongzl replied:
nice catch haha thanks

Ongzl committed:
A-CodingStandard, add javadoc to todo toString

Ongzl committed:
Merge branch 'A-Coding_Standard'

felissaf commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,31 @@
+import duke.util.Parser;
+import duke.util.Ui;
+
 public class Duke {
```
I believe we need to write javadocs for all public classes? This applies for all your other public classes!:smile:

felissaf commented on src/main/java/duke/util/List.java:
```diff
@@ -0,0 +1,149 @@
+import java.time.format.DateTimeParseException;
+import java.util.ArrayList;
+
+public class List {
```
Maybe you can consider giving this class a more specific name, such as TaskList? I feel that List might be too general for this class as this class is specialised for tasklist only.

felissaf commented on src/main/java/duke/util/List.java:
```diff
@@ -0,0 +1,149 @@
+        } catch (DukeException | DateTimeParseException e) {
+            System.out.println("Please input your date and time in dd/mm/yyyy hhmm format, eg. 28/08/2021 2359");
+        }
+
```
I believe that this empty line is not necessary?:smile:

Xiaoyunnn commented on src/main/java/duke/util/List.java:
```diff
@@ -0,0 +1,149 @@
+
+public class List {
+    //List that holds all the Tasks.
+    private final ArrayList<Task> list;
```
According to the [coding standards](https://se-education.org/guides/conventions/java/intermediate.html#naming), plural forms should be used on names representing a collection of objects. So maybe you can change `list` to `tasks`?

Xiaoyunnn commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,99 @@
+                }
+
+                switch (input) {
+                case "bye":
+                case "Bye":
+                case "BYE":
```
According to the [coding standards](https://se-education.org/guides/conventions/java/intermediate.html#layout), the explicit `// Fallthrough` comment should be included whenever there is a case statement without a break statement. So perhaps you can add `// Fallthrough` at the end of line 79?

Xiaoyunnn commented on src/main/java/duke/util/Ui.java:
```diff
@@ -0,0 +1,20 @@
+    /**
+     * Welcome statement for new user.
+     */
+    public void welcome() {
```
Perhaps your method name could be more descriptive here? For example, change `welcome` to `displayWelcomeMsg`.

Agentum07 replied:
Seconded.

Xiaoyunnn reviewed (commented):
LGTM! Just some nits to fix for your JavaDoc and to better align with the coding standards! :D

Ongzl committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Ongzl committed:
A-Gradle, add build gradle

Ongzl committed:
A-CheckStyle, add checkstyle.xml and suppressions.xml to config/checkstyle

Ongzl committed:
Level-10, add javafx to build.gradle

Ongzl committed:
Level-10, add javafx and changes to TaskList.java behaviour to support GUI

Ongzl committed:
Add/edit javadocs

Ongzl committed:
Add/edit javadocs

Ongzl committed:
Add ability to close GUI when bye is sent to MainWindow.java and fixed add event bug, also clear start mtd from duke

Agentum07 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,45 @@
+
+
+
+public class Duke{
```
I think you should have a comment explaining the purpose of the class

Agentum07 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,45 @@
+    }
+
+    /**
+     * Gets response from Duke with the input being the prompt.
```
I think your javadoc could be a little less redundant.
something like
```
   /**
     * Gets response from Duke.
     *
     * @param input Prompt entered by user.
     * @return Duke's reponse as a String.
     */

Agentum07 commented on src/main/java/Main.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+
+}
```
I think you should add a new blank line here.

Agentum07 commented on src/main/java/MainWindow.java:
```diff
@@ -0,0 +1,55 @@
+        scrollPane.vvalueProperty().bind(dialogContainer.heightProperty());
+    }
+
+    public void setDuke(Duke d) {
```
Maybe have a better parameter name than `d`?

Agentum07 commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,87 @@
+                }
+            }
+        } catch (DukeException e) {
+            return "OOPS!!! I'm sorry, but I don't know what that means :-(";
```
Maybe create a dukeException with this output?

Agentum07 commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,160 @@
+    //List that holds all the Tasks.
+    private final ArrayList<Task> list;
+    //Storage to load the list and write to list.
+    private final Storage store;
```
I think have a more descriptive name than `store`

Agentum07 commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,160 @@
+     * Adds ToDo Task to list.
+     *
+     * @param text Description of ToDo Task.
+     */
```
Maybe add a `@return` statement in the javadoc?

Agentum07 commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,160 @@
+     * Sets Task at list index to done .
+     *
+     * @param index Position of Task to be set to done.
+     * @return
```
Incomplete javadoc?

Agentum07 commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,160 @@
+     * @param index Position of Task to be set to done.
+     * @return
+     */
+    public String setIndexDone(int index) {// starts from 1
```
Maybe add the comment on line 91 to increase readability?
Same for the comment on line 91.

Agentum07 commented on src/main/java/duke/util/TaskList.java:
```diff
@@ -0,0 +1,160 @@
+     *
+     * @param index Index of task to be deleted in list.
+     */
+    public String deleteTask(int index) {//starts from 1
```
Same here.

Agentum07 reviewed (commented):
Overall code looks good.
Slight issues with readability due to comments and javadocs.

Good job!

Ongzl committed:
fix checkstyling

Ongzl committed:
A-Assertions
Add assert statement to Deadline.java and Event.java

There are currently checks to make sure that the time variable of
the 2 class are not null

But by adding the assert statement to make sure that the variables are
not null will ensure that there
is no chance of a nullPointerException happening when the toString()
method is used

Add assert statement to check that the time variable in the 2 classes
are not null during construction of the Task object.

Ongzl committed:
A-CodeQuality
Fix Parser.java catch DukeException where it returns an arbitrary string
 instead of getting the message from the exception, also
check through for any code quality violations and attempt to
mitigate/fix.

Previously added overloaded constructor for DukeException to create DE
with custom messages but did not change Parser.java catch behavior to
accommodate change

Change catch statement to get message from exception instead of return
arbitrary error message.

Ongzl committed:
Merge pull request #1 from Ongzl/branch-A-Assertions

A-Assertions

Ongzl committed:
Merge pull request #2 from Ongzl/branch-A-CodeQuality

Branch a code quality

Ongzl committed:
C-Update
Add edit method to TaskList.java as well as support of "edit" keyword in
 Parser.java, add setDescription, setDate, setTime for Task

 Add new feature to Duke, allow user to edit their existing tasks to
 change description, date or time with cmd "edit [task index number]
 /[task type eg.description, date or time] [what to replace with]. Task
 now include Date and Time variable so Event and Deadline can set date
 and time easily while it should not matter to the ToDo class as it is
 never initialized.

Ongzl committed:
Fix bug where edit description only changed the first word

Ongzl committed:
Add and edit javadocs to TaskList.java

Ongzl committed:
Merge pull request #3 from Ongzl/branch-C-Update

Branch c update

Ongzl committed:
A-BetterGui
change profile picture of user and duke, used GUI from @GeNiaaz

Ongzl committed:
Refactor some java files into relevant packages and updated the UI to
better fit the JavaFX GUI

Ongzl committed:
Add more content to README.md, add \n to return string on findTask in
TaskList.java

Ongzl committed:
Changes Duke.kar to duke.jar in README.md

Ongzl committed:
Edit Javadocs for Deadline, DialogBox, Duke, Event, MainWindow and Ui

Ongzl committed:
Update Readme.md removed <br> from code blocks
